### PR TITLE
LOVE-1229 Create simple blueprint for Azure ARM

### PR DESCRIPTION
**Pull request by sffortytwo**
## Definition of Done

This is the definition of done for (new and modified) blueprints to be accepted into the [central XebiaLabs blueprints repository](https://github.com/xebialabs/blueprints).

### Value

This section describes how to determine whether a blueprint has enough value to be included in the central XebiaLabs blueprints repository.

- [ ] The blueprint applies to a focused use case.
- [ ] The blueprint uses at least one XebiaLabs product.
- [ ] The blueprint depends on as few other products as possible. **N.B.:** This is part of the definition of done to keep the blueprint focused and applicable to as many users as possible. A blueprint that depends on one specific CI tool **and** one specific Java EE application server **and** one specific ticketing system can only be used by users that have all three products.
- [ ] The blueprint can be used with content supplied by the user but also includes demo content. A blueprint that does not work with content supplied by the user is a demo or a showcase and does not help the user with their own work.

### Technical

This section describes how to determine whether the blueprint has the right technical quality to be included in the central XebiaLabs blueprints repository.

- [ ] The branch from which the PR is created is up-to-date with the `development` branch.
- [ ] The blueprint contains a `README.md` file at the root of the blueprint folder that describes the blueprint, using the README template in the `.github` folder of this repository.
- [ ] The blueprint generates a `xebialabs/USAGE.md` file which, at a minimum, explains how to use the blueprint after it is generated (like adding any missing steps, creating accounts, setting up Docker containers, applying the YAML using `xl` CLI, running release. etc.). **N.B.:** Do not use this document to describe how to instantiate the blueprint. It will only be available to the user *after* the blueprint has been instantiated.
- [ ] If the blueprint depends on external products that can be started as a Docker container, the blueprint generates a `docker/docker-compose.yml` file so that it can be tested without having to manually install those products. Do not use this Docker Compose file to start XL Deploy, XL Release, Docker or Kubernetes.
- [ ] The blueprint does not contain sensitive information such passwords, tokens, credentials or licenses.
- [ ] The blueprint uses `secret: true` in the `blueprint.yaml` parameter definition for question that ask for sensitive information.
- [ ] The blueprint does not contain the files `xebialabs/.gitignore`, `xebialabs/values.xlvals` and `xebialabs/secrets.xlvals` and does not refer to them from the files section of the `blueprint.yaml` file. These files will be generated when the blueprint is instantiated. To generate the `xebialabs/values.xlvals` file, use the `saveInXlVals: true` directive in the parameters section of the blueprint. To generate the `xebialabs/secrets.xlvals` file, use the `secret: true` directive .
- [ ] The blueprint does not define parameters for trivial things like phase names, task names, folder names etc. Ask questions that add value and be opinionated where possible. For example, ask for a project name and derive folder names, task names etc from that.
- [ ] Folder and file names must use [kebab-case](http://wiki.c2.com/?KebabCase), for example: `aws/sample-app-demo`, `xld-environment.yaml`

### Review and testing

This section describes how to determine whether the blueprint has been reviewed and tested well enough to be included in the central XebiaLabs blueprints repository.

- [ ] Unit test is added in a `__test__` folder for each blueprint. Refer the `CONTRIBUTING.md` file at the root of of this repository for more details.
- [ ] The Travis CI for the PR is green
- [ ] The blueprint has been reviewed by someone else in the team.
- [ ] Both README and USAGE files have been reviewed by a technical writer and a product marketing manager.
- [ ] The blueprint has been manually tested with the `docker/docker-compose.yml` that is generated as part of it.
- [ ] The blueprint works on Linux, Mac and Windows.

### CODE REVIEW

File: azure/basic-arm-template/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+Azure Resource Manager is the deployment and management service for Azure. It provides a consistent management layer that enables you to create, update, and delete resources in your Azure subscription. You can use its access control, auditing, and tagging features to secure and organize your resources after deployment.
+
+Use this blueprint to provision a single VM using an ARM template.
+
+## Before you get started
+
+If you're new to XebiaLabs blueprints, check out:
+
+* [Get started with DevOps as Code](https://docs.xebialabs.com/xl-platform/concept/get-started-with-devops-as-code.html)
```
Comment (sebastian-ziegler) [inline azure/basic-arm-template/README.md]: This URL and the one in the next line doesn't work. I think now is not under `xl-platform` but `xl-release`

File: azure/basic-arm-template/USAGE-skeleton.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Blueprint skeleton
+
+Now that you've created your blueprint skeleton, there are a few more steps:
+
+1. If your blueprint is related to one of technologies (e.g. `aws`), move it to that directory before committing it
+2. If your blueprint is for a whole new technology, create that directory, then move your blueprint to that directory before committing it
+3. When you're done reading this file, delete it, as it should not be part of your blueprint
+
+The files in this skeleton make up a basic blueprint with the most common elements. For more detail, refer to the online documentation:
+* [Blueprint YAML format](https://docs.xebialabs.com/xl-platform/concept/blueprint-yaml-format.html)
```
Comment (sebastian-ziegler) [inline azure/basic-arm-template/USAGE-skeleton.md]: This URL doesn't work

File: azure/basic-arm-template/blueprint.yaml [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    includeIf: GenerateDockerComposeSetup
+    parameterOverrides:
+    - name: UseXLDeploy
+      value: true
+    - name: XLDeployPort
+      value: 4516
+    - name: UseXLRelease
+      value: false
+    - name: XLVersion
+      value: "9.0.0"
```
Comment (sebastian-ziegler) [inline azure/basic-arm-template/blueprint.yaml]: This needs to be changed to `9.0`

Comment (sebastian-ziegler) [reply azure/basic-arm-template/blueprint.yaml]: Or `9.1` if arm plugin needs to be updated

File: azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      - name: {{$app}}-template
+        type: azure.arm.Template
+        deploymentMode: Incremental
+        resourceGroupName: {{.ResourceGroup}}
+        location: {{.ResourceGroupLocation}}
+        deploymentName: {{.AppName}}-deployment
+        file: !file "../arm/template.json"
+        inputVariables:
+          adminUsername: {{.AdminUsername}}
+          adminPassword: !value AdminPassword
```
Comment (sebastian-ziegler) [inline azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl]: This password is been requested as a SecretInput but then is being put here as plain text. Is not better to use an encrypted dictionary in this case??

File: azure/basic-arm-template/blueprint.yaml [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+  includeAfter:
+  - blueprint: xl-devops-platform
+    includeIf: GenerateDockerComposeSetup
+    parameterOverrides:
+    - name: UseXLDeploy
+      value: true
+    - name: XLDeployPort
+      value: 4516
+    - name: UseXLRelease
```
Comment (sebastian-ziegler) [inline azure/basic-arm-template/blueprint.yaml]: At the end there are 2 files generated `xlr-pipeline-ci-cd.yaml` and `xlr-pipeline-destroy.yaml`. This 2 files are dependant on XLR, why is this in false?

File: azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl (hunk 4 above)
Comment (sffortytwo) [reply azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl]: The password comes from the `secrets.xlvals` file. I think this is an accepted standard. I am working on a blueprint that showcases encrypted dictionaries though.

File: azure/basic-arm-template/blueprint.yaml (hunk 5 above)
Comment (sffortytwo) [reply azure/basic-arm-template/blueprint.yaml]: Yeah, that's a mistake. I've marked it as true

File: azure/basic-arm-template/USAGE-skeleton.md (hunk 2 above)
Comment (sffortytwo) [reply azure/basic-arm-template/USAGE-skeleton.md]: Fixed both URLs

File: azure/basic-arm-template/README.md (hunk 1 above)
Comment (sffortytwo) [reply azure/basic-arm-template/README.md]: Fixed both URLs

File: azure/basic-arm-template/USAGE-skeleton.md (hunk 2 above)
Comment (sffortytwo) [reply azure/basic-arm-template/USAGE-skeleton.md]: Fixed

File: azure/basic-arm-template/blueprint.yaml (hunk 3 above)
Comment (sffortytwo) [reply azure/basic-arm-template/blueprint.yaml]: What's the next version that's being released? 9.5? I'll find out

Comment (sffortytwo) [reply azure/basic-arm-template/blueprint.yaml]: But I chatted with Deepu and there's definitely a consensus to have version numbers always be x.y.z, even in there are two zeroes

Comment (sffortytwo) [reply azure/basic-arm-template/blueprint.yaml]: I've updated it to 9.5.0 (the next release)

File: azure/basic-arm-template/blueprint.yaml [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      value: false
+
+  files:
+  - path: xebialabs.yaml
+  - path: xebialabs/USAGE.md.tmpl
+  - path: xebialabs/xld-apps.yaml.tmpl
+  - path: xebialabs/xld-infra-env.yaml.tmpl
+  - path: xebialabs/xlr-pipeline-ci-cd.yaml.tmpl
+  - path: xebialabs/xlr-pipeline-destroy.yaml.tmpl
+  - path: arm/template.json
```
Comment (deepu105) [inline azure/basic-arm-template/blueprint.yaml]: create this only if `TemplateFile` value hasnt changed from default

File: azure/basic-arm-template/xebialabs/USAGE.md.tmpl [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    xl apply -f xebialabs.yaml
+    ```
+
+2. Go to XL Release, look for the "{{$app}}-ci-cd" template, and start a new release from it.
+
+## Minimum required versions
+
+This blueprint version requires at least the following versions of the specified tools to work properly:
+
+- XL Deploy: Version 9.0.0
```
Comment (deepu105) [inline azure/basic-arm-template/xebialabs/USAGE.md.tmpl]: versions should be 9.5

File: azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    - name: '1.0.0'
+      type: udm.DeploymentPackage
+      deployables:
+      - name: {{$app}}-template
+        type: azure.arm.Template
+        deploymentMode: Incremental
+        resourceGroupName: {{.ResourceGroup}}
+        location: {{.ResourceGroupLocation}}
+        deploymentName: {{.AppName}}-deployment
+        file: !file "../arm/template.json"
```
Comment (deepu105) [inline azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl]: Shouldnt this be `{{TemplateFile}}` else what is the point of that question?

File: azure/basic-arm-template/README.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
+# Basic Arm Template
```
[the remark this answers is not in the document]
Comment (sffortytwo) [reply azure/basic-arm-template/README.md]: Fixed

File: azure/basic-arm-template/USAGE-skeleton.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# Blueprint skeleton
```
[the remark this answers is not in the document]
Comment (sffortytwo) [reply azure/basic-arm-template/USAGE-skeleton.md]: It's a leftover from the blueprint-making blueprint. Removed.

File: azure/basic-arm-template/xebialabs/USAGE.md.tmpl (hunk 7 above)
Comment (sffortytwo) [reply azure/basic-arm-template/xebialabs/USAGE.md.tmpl]: Fixed version

File: azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl (hunk 8 above)
Comment (sffortytwo) [reply azure/basic-arm-template/xebialabs/xld-apps.yaml.tmpl]: Yes, I missed that. Fixed.

File: azure/basic-arm-template/blueprint.yaml (hunk 6 above)
Comment (sffortytwo) [reply azure/basic-arm-template/blueprint.yaml]: Fixed

**Pull request merged**